File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res = nums[0];
        int currMin = 1, currMax = 1;
        for(const int& n: nums){
            int temp = n*currMax;
            currMax = max(max(temp, n*currMin), n);
            currMin = min(min(temp, n*currMin), n);
            res = max(res, currMax);
        }
        return res; 
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1]*n
        i = 1
        while i < n:
            res[i] = nums[i-1]*res[i-1]
            i+=1
        i = n-2
        temp = nums[n-1]
        while i >= 0:
            res[i] *= temp
            temp *= nums[i]
            i-=1
        return res
